pillarbox-player/src/main/java/ch/srgssr/pillarbox/player/tracker/ChaptersTracker.kt

```diff
@@ -0,0 +1,124 @@
+import kotlin.time.Duration.Companion.milliseconds
+
+/**
+ * Track chapter changes during playback.
+ */
+internal class ChaptersTracker(
+    private val pillarboxExoPlayer: PillarboxExoPlayer
+) : CurrentMediaItemPillarboxDataTracker.Callback {
+
+    private val listPlayerMessage = mutableListOf<PlayerMessage>()
+    private var listChapter = emptyList<Chapter>()
+    private val listener = Listener()
+
+    private var lastChapter: Chapter? = pillarboxExoPlayer.getChapterAtPosition()
+        set(value) {
+            if (field != value) {
+                field = value
+                pillarboxExoPlayer.notifyCurrentChapterChanged(field)
+            }
+        }
+
+    override fun onPillarboxDataChanged(mediaItem: MediaItem?, data: PillarboxData?) {
+        clearPlayerMessage()
+        lastChapter = pillarboxExoPlayer.getChapterAtPosition()
+        pillarboxExoPlayer.removeListener(listener)
+        if (data == null || mediaItem == null) return
+        listChapter = mediaItem.pillarboxData.chapters
+        pillarboxExoPlayer.addListener(listener)
+        createMessages()
+    }
+
+    private fun createMessages() {
+        val messageHandler = PlayerMessage.Target { messageType, message ->
+            val chapter = message as Chapter
+            when (messageType) {
+                TYPE_ENTER -> {
+                    if (chapter != lastChapter) {
+                        lastChapter = chapter
+                        Log.i(TAG, "Enter chapter $chapter")
+                    }
+                }
+
+                TYPE_EXIT -> {
+                    Log.i(TAG, "Exit chapter $chapter")
+                    lastChapter = null
+                }
+            }
+        }
+
+        listChapter.forEach {
+            val messageEnter = pillarboxExoPlayer.createMessage(messageHandler).apply {
+                deleteAfterDelivery = false
+                looper = pillarboxExoPlayer.applicationLooper
+                payload = it
+                setPosition(it.start)
+                type = TYPE_ENTER
+            }
+            messageEnter.send()
+            listPlayerMessage.add(messageEnter)
+
+            val messageExit = pillarboxExoPlayer.createMessage(messageHandler).apply {
+                deleteAfterDelivery = false
+                looper = pillarboxExoPlayer.applicationLooper
+                payload = it
+                setPosition(it.end)
+                type = TYPE_EXIT
+            }
+            messageExit.send()
+            listPlayerMessage.add(messageExit)
+        }
+    }
+
+    private fun clearPlayerMessage() {
+        listPlayerMessage.forEach {
+            it.cancel()
+        }
+        listPlayerMessage.clear()
+    }
+
+    private inner class Listener : Player.Listener {
+        override fun onEvents(player: Player, events: Player.Events) {
+            if (events.containsAny(
+                    Player.EVENT_POSITION_DISCONTINUITY
+                )
+            ) {
```

Comment:
Why not override `onPositionDiscontinuity` directly?
```suggestion
        override fun onPositionDiscontinuity(oldPosition: Player.PositionInfo, newPosition: Player.PositionInfo, reason: Int) {
```

And where you use the `player` argument, you can use `pillarboxExoPlayer` instead.

Reply:
Because why not, perhaps during dev more events was needed. But I totally agree right now. But we have to handle item transition as well here.